docs/modules/ROOT/pages/design/cloudevents.adoc

```diff
@@ -0,0 +1,125 @@
+= Cloud Event Delivery Pipeline
+== Overview
+
+KubeArchive uses Cloud Events that are generated by a Knative Eventing ApiServerSource to discover when resources on
+the kubernetes cluster have changed. These Cloud Events contain the json representation of the Kubernetes resource that
+has changed. The KubeArchive Sink (sink) receives these messages and determines, using CEL expression from the
+KubeArchiveConfig, if the data contained in the Cloud Event should be written to the database and potentially deleted
+from the cluster. While the sink is processing a Cloud Event, it is possible at many points for an error to occur,
+especially when the sink has to make a request of the network, particularly, to the Kubernetes API or the database. This
+page proposes how KubeArchive set up its Cloud Event delivery pipeline and how it should consume Cloud Events to achieve
+the following goals:
+
+* KubeArchive receives all Cloud Events from the Knative ApiServerSource
+* Cloud Events sent to KubeArchive can be replayed
+* Cloud Events received by KubeArchive, that cannot be processed successfully, are reported back to the sender as
+  delivery failure
+* Cloud Events sent to the sink will attempt to be redelivered if the initial delivery fails
+* Cloud Events that could not be delivered are replayed to KubeArchive automatically when KubeArchive becomes available
+  again
+
+== Design
+
+KubeArchive needs to use the functionality provided by Knative Eventing, Brokers and Triggers as well as features
+provided by the Cloud Events Specification to achieve an event driven architecture, where events are not lost and events
+can be replayed safely without worries about side effects.
```

Comment:
```suggestion
We decided to use an Event Driven architecture for KubeArchive. Reusing Knative Eventing's components the system can be fault tolerant and scale well with multiple event consumers.
```

Reply:
Made the change. I tweaked your suggestion slightly